pandas/_libs/interval.pyi

```diff
@@ -0,0 +1,134 @@
+    @property
+    def is_empty(self: IntervalMixinProtocol) -> bool: ...
+    def _check_closed_matches(self, other: IntervalMixin, name: str = ...): ...
+
+class Interval(IntervalMixin, Generic[OrderableT]):
+    @property
+    def left(self: Interval[OrderableT]) -> OrderableT: ...
+    @property
+    def right(self: Interval[OrderableT]) -> OrderableT: ...
+    @property
+    def closed(self) -> str: ...
+    def __init__(
+        self,
+        left: OrderableT,
+        right: OrderableT,
+        closed: Union[str, Literal["left", "right", "both", "neither"]] = ...,
+    ): ...
+    length: _LengthDescriptor
+    def __hash__(self) -> int: ...
+    @overload
+    def __contains__(self: Interval[OrderableTimesT], OrderableTimesT) -> bool: ...
+    @overload
+    def __contains__(self: Interval[int], key: Union[int, float]) -> bool: ...
+    @overload
+    def __contains__(self: Interval[float], key: Union[int, float]) -> bool: ...
+    def __repr__(self) -> str: ...
+    def __str__(self) -> str: ...
+    @overload
+    def __add__(
+        self: Interval[OrderableTimesT], y: Timedelta
+    ) -> Interval[OrderableTimesT]: ...
+    @overload
+    def __add__(self: Interval[int], y: int) -> Interval[int]: ...
+    @overload
+    def __add__(self: Interval[int], y: float) -> Interval[float]: ...
+    @overload
+    def __add__(self: Interval[float], y: Union[int, float]) -> Interval[float]: ...
+    @overload
+    def __sub__(
+        self: Interval[OrderableTimesT], y: Timedelta
+    ) -> Interval[OrderableTimesT]: ...
+    @overload
+    def __sub__(self: Interval[int], y: int) -> Interval[int]: ...
+    @overload
+    def __sub__(self: Interval[int], y: float) -> Interval[float]: ...
+    @overload
+    def __sub__(self: Interval[float], y: Union[int, float]) -> Interval[float]: ...
+    @overload
+    def __mul__(self: Interval[int], y: int) -> Interval[int]: ...
+    @overload
+    def __mul__(self: Interval[int], y: float) -> Interval[float]: ...
+    @overload
+    def __mul__(self: Interval[float], y: Union[int, float]) -> Interval[float]: ...
+    @overload
+    def __truediv__(self: Interval[int], y: int) -> Interval[int]: ...
+    @overload
+    def __truediv__(self: Interval[int], y: float) -> Interval[float]: ...
+    @overload
+    def __truediv__(self: Interval[float], y: Union[int, float]) -> Interval[float]: ...
+    @overload
+    def __floordiv__(self: Interval[int], y: int) -> Interval[int]: ...
+    @overload
+    def __floordiv__(self: Interval[int], y: float) -> Interval[float]: ...
+    @overload
+    def __floordiv__(
+        self: Interval[float], y: Union[int, float]
+    ) -> Interval[float]: ...
+    def overlaps(self: Interval[OrderableT], other: Interval[OrderableT]) -> bool: ...
+
+def intervals_to_interval_bounds(intervals: np.ndarray, validate_closed: int = ...): ...
+
+class IntervalTree(IntervalMixin):
+    def __init__(
+        self,
+        left: np.ndarray,
```

Comment:
You can make this class generic to keep track of the dtype

Reply:
Probably should be generic, because `IntervalMixin` needs to be generic (for length/mid).

Reply:
Not sure that is necessary - we're OK without it

Reply:
I see you found a way to avoid `IntervalMixin` being generic :)

Since the implementation [appears](https://github.com/pandas-dev/pandas/blob/ea0184e339b7dbf27eec53e6c01c6514bda1fc83/pandas/_libs/intervaltree.pxi.in#L61) to allow ndarrays of different types, it should be fine not to track the dtype of the ndarray.

Reply:
later, you indicated to remove the generic, so resolving this